Deploy docs to gh-pages from GitHub CI

nilsvu opened the pull request:
## Proposed changes

Switches the docs deployment from Travis to GitHub CI.

Requires a GitHub personal access token named `GH_PAGES_DEPLOY_TOKEN` to be added to the repository's "Settings" > "Secrets" page. [This page](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line) describes how to create such a token. It must include the `public_repo` permissions. Once [an issue](https://github.community/t5/GitHub-Actions/Github-action-not-triggering-gh-pages-upon-push/td-p/26869) with GitHub Actions is resolved, we can just use the automatic `secrets.GITHUB_TOKEN` instead, so no personal access token is required anymore.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [x] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Deploy docs to gh-pages from GitHub CI

nilsdeppe commented:
Okay, key should be added. I only enabled the `public_repo` permission, no admin rights or anything. How can we test this?

nilsdeppe reviewed (commented):
This looks good! One concern I have is that we will create a large repo history if we keep all past dox commits. This happened to us previously and that's why the Travis way the docs were deployed always deleted the previous dox commits. Is that possible to do here?

nilsvu commented:
- I think the only way to test this is on a fork. I did on mine, and if you want to do another test you can just push the commit to your fork’s develop branch. It will trigger the deployment to nilsdeppe.github.com/spectre. I’m not sure how the permissions work out though, and I can’t think of a good way to test it. You can always create another access token and edit the one the settings.
- The action I’m currently using doesn’t support orphaned branches, but there’s another one that does. I’ll give it a try.

nilsvu committed:
Add .nojekyll file to bypass Jekyll on gh-pages

We don't need GitHub Pages to run Jekyll over the html docs. So this
commit disables it to avoid potential problems (e.g. Jekyll discards
directories that start with an underscore).

nilsvu commented:
@nilsdeppe I pushed a fixup that makes the action discard the gh-pages history on deployment, and another that handles the `CNAME` file for the domain name. Another new commit adds a `.nokeyll` file. Please take a look.

nilsdeppe commented on .github/workflows/Tests.yaml:
```diff
@@ -174,6 +174,11 @@ jobs:
+        if: github.repository == 'sxs-collaboration/spectre'
+        working-directory: ./docs-html
+        run: |
+          echo "spectre-code-.org" >> CNAME
```
Just one `>` so that it overwrites if the file happens to exist

nilsdeppe reviewed (commented):
LGTM! Just one minor comment. I'm happy for you to squash. The `.nojekyll` thing is awesome! Nice find :)

nilsdeppe commented on .github/workflows/Tests.yaml:
```diff
@@ -154,6 +161,38 @@ jobs:
+        if: github.repository == 'sxs-collaboration/spectre'
+        working-directory: ./docs-html
+        run: |
+          echo "spectre-code-.org" > CNAME
```
only one `-` in `spectre-code.org`

nilsvu replied:
Copied this from https://github.com/sxs-collaboration/spectre/commit/5e933a48d6e063cac87f2f285a9cb2696208e8bd which has the additional `-`. Do you know why?

nilsdeppe replied:
Yea, I do know why... Something in the settings for the repo was wrong. That's fixed now.

nilsleiffischer replied:
Was it wrong on Travis?

I fixed and squashed this.

nilsvu replied:
Was it wrong on Travis?

I fixed and squashed this.

nilsdeppe replied:
It was never used with Travis. That was something GitHub added recently while Larry and I were looking at changing where the documentation is hosted from.

nilsdeppe reviewed (commented):
Review with @kidder please squash the one small typo fix. Looks good!
